File: tests/conformance/v1/contract.test.ts
/**
 * Conformance test suite — Contract v1 (api-orchestrator side)
 *
 * These tests validate that responses from engine-compute match
 * the shapes defined in docs/contracts/v1/openapi.yaml.
 *
 * To run against a live engine:  ENGINE_URL=http://localhost:8080 npm run test:conformance
 * To run in schema-only mode:    npm run test:conformance (validates fixtures only)
 */

import * as fs from "fs";
import * as path from "path";

const FIXTURES_DIR = path.join(import.meta.dirname, "fixtures");
const ENGINE_URL = process.env.ENGINE_URL ?? null;

function loadFixture(name: string): unknown {
  return JSON.parse(
    fs.readFileSync(path.join(FIXTURES_DIR, name), "utf-8")
  );
}

// --- Schema shape assertions ---

function assertContractVersion(obj: Record<string, unknown>, label: string): void {
  if (obj["_contract_version"] !== "v1") {
    throw new Error(
      `${label}: expected _contract_version "v1", got ${JSON.stringify(obj["_contract_version"])}`
    );
  }
}

function assertObservabilityContext(obj: Record<string, unknown>, label: string): void {
  const ctx = obj["_observability"] as Record<string, unknown> | undefined;
  if (!ctx || typeof ctx !== "object") {
    throw new Error(`${label}: '_observability' must be an object`);
  }
  for (const field of ["request_id", "trace_id", "correlation_id"]) {
    if (typeof ctx[field] !== "string" || (ctx[field] as string).length === 0) {
      throw new Error(`${label}: '_observability.${field}' must be a non-empty string`);
    }
  }
}

function assertStorageMetadata(obj: Record<string, unknown>, label: string): void {
  const storage = obj["_storage"] as Record<string, unknown> | undefined;
  if (!storage || typeof storage !== "object") {
    throw new Error(`${label}: '_storage' must be an object`);
  }
  const validOwners = ["redis", "postgres", "ephemeral"];
  const validDurability = ["volatile", "persistent"];
  const validConsistency = ["eventual", "strong", "not_applicable"];
  if (!validOwners.includes(storage["owner"] as string)) {
    throw new Error(`${label}: '_storage.owner' invalid`);
  }
  if (!validDurability.includes(storage["durability"] as string)) {
    throw new Error(`${label}: '_storage.durability' invalid`);
  }
  if (!validConsistency.includes(storage["consistency"] as string)) {
    throw new Error(`${label}: '_storage.consistency' invalid`);
  }
  if (typeof storage["invalidation_scope"] !== "string") {
    throw new Error(`${label}: '_storage.invalidation_scope' must be a string`);
  }
}

function assertMoveResponse(obj: Record<string, unknown>): void {
  assertContractVersion(obj, "MoveResponse");
  assertObservabilityContext(obj, "MoveResponse");
  assertStorageMetadata(obj, "MoveResponse");
  if (typeof obj["move"] !== "string") {
    throw new Error(`MoveResponse: 'move' must be a string`);
  }
}

function assertTimeoutResponse(obj: Record<string, unknown>): void {
  assertContractVersion(obj, "TimeoutResponse");
  assertObservabilityContext(obj, "TimeoutResponse");
  if (obj["error"] !== "timeout") {
    throw new Error(`TimeoutResponse: 'error' must be "timeout"`);
  }
  if (!("partial_result" in obj)) {
    throw new Error(`TimeoutResponse: 'partial_result' field must be present (nullable)`);
  }
  if (typeof obj["depth_reached"] !== "number") {
    throw new Error(`TimeoutResponse: 'depth_reached' must be a number`);
  }
  if (typeof obj["elapsed_ms"] !== "number") {
    throw new Error(`TimeoutResponse: 'elapsed_ms' must be a number`);
  }
}

function assertErrorResponse(obj: Record<string, unknown>): void {
  assertContractVersion(obj, "ErrorResponse");
  assertObservabilityContext(obj, "ErrorResponse");
  const validErrors = ["invalid_input", "unavailable", "timeout", "internal_error", "dependency_error", "rate_limited"];
  const validStatusCodes = [400, 429, 500, 503, 504];
  if (!validErrors.includes(obj["error"] as string)) {
    throw new Error(
      `ErrorResponse: 'error' must be one of ${validErrors.join(", ")}, got ${obj["error"]}`
    );
  }
  if (!validStatusCodes.includes(obj["status_code"] as number)) {
    throw new Error(`ErrorResponse: 'status_code' must be one of ${validStatusCodes.join(", ")}`);
  }
  if (typeof obj["retryable"] !== "boolean") {
    throw new Error(`ErrorResponse: 'retryable' must be a boolean`);
  }
}

function assertEndgameResponse(obj: Record<string, unknown>): void {
  assertContractVersion(obj, "EndgameResponse");
  if (typeof obj["available"] !== "boolean") {
    throw new Error(`EndgameResponse: 'available' must be a boolean`);
  }
  const validProviderStates = ["hit", "miss", "timeout", "unavailable", "error", "quarantined"];
  if (!validProviderStates.includes(obj["provider_status"] as string)) {
    throw new Error(`EndgameResponse: invalid provider_status ${obj["provider_status"]}`);
  }
  const validFallbackActions = ["none", "search", "opening", "fail_closed"];
  if (!validFallbackActions.includes(obj["fallback_action"] as string)) {
    throw new Error(`EndgameResponse: invalid fallback_action ${obj["fallback_action"]}`);
  }
  if (typeof obj["retryable"] !== "boolean") {
    throw new Error(`EndgameResponse: 'retryable' must be a boolean`);
  }
  if (obj["provider_status"] === "quarantined") {
    const quarantine = obj["quarantine"] as Record<string, unknown> | null;
    if (!quarantine || quarantine["active"] !== true) {
      throw new Error(`EndgameResponse: quarantined status requires active quarantine metadata`);
    }
  }
}

// --- Fixture conformance tests ---

function testMoveResponseFixture(): void {
  const fixture = loadFixture("move_response_valid.json") as Record<string, unknown>;
  assertMoveResponse(fixture);
  console.log("  PASS: MoveResponse fixture matches schema");
}

function testTimeoutResponseFixture(): void {
  const fixture = loadFixture("timeout_response.json") as Record<string, unknown>;
  assertTimeoutResponse(fixture);
  console.log("  PASS: TimeoutResponse fixture matches schema");
}

function testErrorUnavailableFixture(): void {
  const fixture = loadFixture("error_unavailable.json") as Record<string, unknown>;
  assertErrorResponse(fixture);
  if (fixture["error"] !== "unavailable") {
    throw new Error(`Expected error=unavailable`);
  }
  console.log("  PASS: ErrorResponse(unavailable) fixture matches schema");
}

function testErrorInvalidInputFixture(): void {
  const fixture = loadFixture("error_invalid_input.json") as Record<string, unknown>;
  assertErrorResponse(fixture);
  if (fixture["error"] !== "invalid_input") {
    throw new Error(`Expected error=invalid_input`);
  }
  console.log("  PASS: ErrorResponse(invalid_input) fixture matches schema");
}

function testEndgameTimeoutFallbackFixture(): void {
  const fixture = loadFixture("endgame_fallback_timeout.json") as Record<string, unknown>;
  assertEndgameResponse(fixture);
  if (fixture["provider_status"] !== "timeout" || fixture["fallback_action"] !== "search") {
    throw new Error(`Expected timeout/search fallback contract`);
  }
  console.log("  PASS: Endgame timeout fallback fixture matches schema");
}

function testEndgameQuarantineFixture(): void {
  const fixture = loadFixture("endgame_quarantined.json") as Record<string, unknown>;
  assertEndgameResponse(fixture);
  if (fixture["provider_status"] !== "quarantined") {
    throw new Error(`Expected provider_status=quarantined`);
  }
  console.log("  PASS: Endgame quarantined fixture matches schema");
}

// --- Live engine tests (skipped if ENGINE_URL not set) ---

async function testLiveMoveSelection(): Promise<void> {
  if (!ENGINE_URL) {
    console.log("  SKIP: testLiveMoveSelection (ENGINE_URL not set)");
    return;
  }
  const request = loadFixture("move_request_valid.json");
  const res = await fetch(`${ENGINE_URL}/v1/move`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });

  if (res.status === 200) {
    const body = (await res.json()) as Record<string, unknown>;
    assertMoveResponse(body);
    console.log(`  PASS: testLiveMoveSelection → move=${body["move"]}`);
  } else if (res.status === 504) {
    const body = (await res.json()) as Record<string, unknown>;
    assertTimeoutResponse(body);
    console.log(`  PASS: testLiveMoveSelection → timeout (depth=${body["depth_reached"]})`);
  } else {
    throw new Error(`Unexpected status ${res.status} from /v1/move`);
  }
}

async function testLiveHealthCheck(): Promise<void> {
  if (!ENGINE_URL) {
    console.log("  SKIP: testLiveHealthCheck (ENGINE_URL not set)");
    return;
  }
  const res = await fetch(`${ENGINE_URL}/v1/health`);
  if (res.status !== 200 && res.status !== 503) {
    throw new Error(`/v1/health returned unexpected status ${res.status}`);
  }
  const body = (await res.json()) as Record<string, unknown>;
  assertContractVersion(body, "HealthResponse");
  const validStatuses = ["ok", "degraded"];
  if (!validStatuses.includes(body["status"] as string)) {
    throw new Error(`HealthResponse: 'status' must be one of ${validStatuses.join(", ")}`);
  }
  console.log(`  PASS: testLiveHealthCheck → status=${body["status"]}`);
}

// --- Runner ---

async function main(): Promise<void> {
  console.log("=== Contract v1 Conformance Tests ===\n");

  let passed = 0;
  let failed = 0;

  const tests: Array<() => void | Promise<void>> = [
    testMoveResponseFixture,
    testTimeoutResponseFixture,
    testErrorUnavailableFixture,
    testErrorInvalidInputFixture,
    testEndgameTimeoutFallbackFixture,
    testEndgameQuarantineFixture,
    testLiveMoveSelection,
    testLiveHealthCheck,
  ];

  for (const test of tests) {
    try {
      await test();
      passed++;
    } catch (err) {
      console.error(`  FAIL: ${test.name} — ${(err as Error).message}`);
      failed++;
    }
  }

  console.log(`\n${passed} passed, ${failed} failed`);
  if (failed > 0) {
    process.exit(1);
  }
}

main().catch((err) => {
  console.error("Conformance runner error:", err);
  process.exit(1);
});

File: tests/conformance/v1/contract_test.rs
/// Conformance test suite — Contract v1 (engine-compute side)
///
/// These tests validate that the engine-compute service produces responses
/// that match the shapes defined in docs/contracts/v1/openapi.yaml.
///
/// Run: cargo test --test conformance
/// With live server: ENGINE_TEST_ADDR=127.0.0.1:8080 cargo test --test conformance

use serde_json::Value;
use std::fs;
use std::path::Path;

const FIXTURES_DIR: &str = "tests/conformance/v1/fixtures";

fn load_fixture(name: &str) -> Value {
    let path = Path::new(FIXTURES_DIR).join(name);
    let content = fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("failed to read fixture {name}: {e}"));
    serde_json::from_str(&content)
        .unwrap_or_else(|e| panic!("failed to parse fixture {name}: {e}"))
}

fn assert_contract_version(obj: &Value, label: &str) {
    let version = obj
        .get("_contract_version")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("{label}: missing _contract_version field"));
    assert_eq!(
        version, "v1",
        "{label}: expected _contract_version=v1, got {version}"
    );
}

fn assert_move_response(obj: &Value) {
    assert_contract_version(obj, "MoveResponse");
    assert_observability_context(obj, "MoveResponse");
    assert_storage_metadata(obj, "MoveResponse");
    assert!(
        obj.get("move").and_then(|v| v.as_str()).is_some(),
        "MoveResponse: 'move' must be a string"
    );
}

fn assert_timeout_response(obj: &Value) {
    assert_contract_version(obj, "TimeoutResponse");
    assert_observability_context(obj, "TimeoutResponse");
    assert_eq!(
        obj.get("error").and_then(|v| v.as_str()),
        Some("timeout"),
        "TimeoutResponse: 'error' must be \"timeout\""
    );
    assert!(
        obj.get("partial_result").is_some(),
        "TimeoutResponse: 'partial_result' field must be present (nullable)"
    );
    assert!(
        obj.get("depth_reached").and_then(|v| v.as_u64()).is_some(),
        "TimeoutResponse: 'depth_reached' must be a number"
    );
    assert!(
        obj.get("elapsed_ms").and_then(|v| v.as_u64()).is_some(),
        "TimeoutResponse: 'elapsed_ms' must be a number"
    );
}

fn assert_error_response(obj: &Value) {
    assert_contract_version(obj, "ErrorResponse");
    assert_observability_context(obj, "ErrorResponse");
    let valid_errors = [
        "invalid_input",
        "unavailable",
        "timeout",
        "internal_error",
        "dependency_error",
        "rate_limited",
    ];
    let valid_status_codes = [400_u64, 429_u64, 500_u64, 503_u64, 504_u64];
    let error_val = obj
        .get("error")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("ErrorResponse: missing 'error' field"));
    assert!(
        valid_errors.contains(&error_val),
        "ErrorResponse: 'error' must be one of {valid_errors:?}, got {error_val}"
    );
    let status_code = obj
        .get("status_code")
        .and_then(|v| v.as_u64())
        .unwrap_or_else(|| panic!("ErrorResponse: missing 'status_code' field"));
    assert!(
        valid_status_codes.contains(&status_code),
        "ErrorResponse: 'status_code' must be one of {valid_status_codes:?}, got {status_code}"
    );
    assert!(
        obj.get("retryable").and_then(|v| v.as_bool()).is_some(),
        "ErrorResponse: 'retryable' must be a boolean"
    );
}

fn assert_observability_context(obj: &Value, label: &str) {
    let obs = obj
        .get("_observability")
        .unwrap_or_else(|| panic!("{label}: missing _observability field"));
    for field in ["request_id", "trace_id", "correlation_id"] {
        let value = obs
            .get(field)
            .and_then(|v| v.as_str())
            .unwrap_or_else(|| panic!("{label}: missing _observability.{field}"));
        assert!(
            !value.is_empty(),
            "{label}: _observability.{field} must be non-empty"
        );
    }
}

fn assert_storage_metadata(obj: &Value, label: &str) {
    let storage = obj
        .get("_storage")
        .unwrap_or_else(|| panic!("{label}: missing _storage field"));
    let owner = storage
        .get("owner")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("{label}: missing _storage.owner"));
    assert!(
        ["redis", "postgres", "ephemeral"].contains(&owner),
        "{label}: invalid _storage.owner {owner}"
    );
    let durability = storage
        .get("durability")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("{label}: missing _storage.durability"));
    assert!(
        ["volatile", "persistent"].contains(&durability),
        "{label}: invalid _storage.durability {durability}"
    );
    let consistency = storage
        .get("consistency")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("{label}: missing _storage.consistency"));
    assert!(
        ["eventual", "strong", "not_applicable"].contains(&consistency),
        "{label}: invalid _storage.consistency {consistency}"
    );
    storage
        .get("invalidation_scope")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("{label}: missing _storage.invalidation_scope"));
}

fn assert_endgame_response(obj: &Value) {
    assert_contract_version(obj, "EndgameResponse");
    assert!(
        obj.get("available").and_then(|v| v.as_bool()).is_some(),
        "EndgameResponse: 'available' must be a boolean"
    );

    let provider_states = ["hit", "miss", "timeout", "unavailable", "error", "quarantined"];
    let provider_status = obj
        .get("provider_status")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("EndgameResponse: missing 'provider_status' field"));
    assert!(
        provider_states.contains(&provider_status),
        "EndgameResponse: invalid provider_status {provider_status}"
    );

    let fallback_actions = ["none", "search", "opening", "fail_closed"];
    let fallback_action = obj
        .get("fallback_action")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("EndgameResponse: missing 'fallback_action' field"));
    assert!(
        fallback_actions.contains(&fallback_action),
        "EndgameResponse: invalid fallback_action {fallback_action}"
    );

    assert!(
        obj.get("retryable").and_then(|v| v.as_bool()).is_some(),
        "EndgameResponse: 'retryable' must be a boolean"
    );

    if provider_status == "quarantined" {
        let quarantine = obj
            .get("quarantine")
            .unwrap_or_else(|| panic!("EndgameResponse: missing 'quarantine' object"));
        assert_eq!(
            quarantine.get("active").and_then(|v| v.as_bool()),
            Some(true),
            "EndgameResponse: quarantine.active must be true when quarantined"
        );
    }
}

// --- Fixture conformance tests ---

#[test]
fn test_move_response_fixture() {
    let fixture = load_fixture("move_response_valid.json");
    assert_move_response(&fixture);
}

#[test]
fn test_timeout_response_fixture() {
    let fixture = load_fixture("timeout_response.json");
    assert_timeout_response(&fixture);
}

#[test]
fn test_error_unavailable_fixture() {
    let fixture = load_fixture("error_unavailable.json");
    assert_error_response(&fixture);
    assert_eq!(
        fixture.get("error").and_then(|v| v.as_str()),
        Some("unavailable")
    );
}

#[test]
fn test_error_invalid_input_fixture() {
    let fixture = load_fixture("error_invalid_input.json");
    assert_error_response(&fixture);
    assert_eq!(
        fixture.get("error").and_then(|v| v.as_str()),
        Some("invalid_input")
    );
}

#[test]
fn test_endgame_timeout_fallback_fixture() {
    let fixture = load_fixture("endgame_fallback_timeout.json");
    assert_endgame_response(&fixture);
    assert_eq!(
        fixture.get("provider_status").and_then(|v| v.as_str()),
        Some("timeout")
    );
    assert_eq!(
        fixture.get("fallback_action").and_then(|v| v.as_str()),
        Some("search")
    );
}

#[test]
fn test_endgame_quarantine_fixture() {
    let fixture = load_fixture("endgame_quarantined.json");
    assert_endgame_response(&fixture);
    assert_eq!(
        fixture.get("provider_status").and_then(|v| v.as_str()),
        Some("quarantined")
    );
}

// --- Live engine tests (skipped if ENGINE_TEST_ADDR not set) ---

#[test]
fn test_live_health_check() {
    let addr = match std::env::var("ENGINE_TEST_ADDR") {
        Ok(a) => a,
        Err(_) => {
            eprintln!("SKIP: ENGINE_TEST_ADDR not set");
            return;
        }
    };

    let url = format!("http://{addr}/v1/health");
    let response = ureq::get(&url)
        .call()
        .unwrap_or_else(|e| panic!("GET /v1/health failed: {e}"));

    let status = response.status();
    assert!(
        status == 200 || status == 503,
        "/v1/health returned unexpected status {status}"
    );

    let body: Value = response
        .into_json()
        .unwrap_or_else(|e| panic!("failed to parse /v1/health response: {e}"));

    assert_contract_version(&body, "HealthResponse");

    let valid_statuses = ["ok", "degraded"];
    let status_val = body
        .get("status")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| panic!("HealthResponse: missing 'status' field"));
    assert!(
        valid_statuses.contains(&status_val),
        "HealthResponse: 'status' must be one of {valid_statuses:?}, got {status_val}"
    );
}

#[test]
fn test_live_move_selection_blitz() {
    let addr = match std::env::var("ENGINE_TEST_ADDR") {
        Ok(a) => a,
        Err(_) => {
            eprintln!("SKIP: ENGINE_TEST_ADDR not set");
            return;
        }
    };

    let url = format!("http://{addr}/v1/move");
    let request_body = serde_json::json!({
        "fen": "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e3 0 1",
        "time_control": "blitz",
        "moves_played": 1,
        "remaining_ms": 180000
    });

    let response = ureq::post(&url)
        .set("Content-Type", "application/json")
        .send_json(request_body)
        .unwrap_or_else(|e| panic!("POST /v1/move failed: {e}"));

    let http_status = response.status();
    let body: Value = response
        .into_json()
        .unwrap_or_else(|e| panic!("failed to parse /v1/move response: {e}"));

    match http_status {
        200 => assert_move_response(&body),
        504 => assert_timeout_response(&body),
        _ => panic!("/v1/move returned unexpected status {http_status}"),
    }
}
